.github/workflows/qns.yml

```diff
@@ -193,22 +201,57 @@ jobs:
                 echo "**$RESULT**"
               } >> "$GROUP.md"
             done
+            jq < "$RUN/grouped.json" ". += { client: \"$CLIENT\", server: \"$SERVER\" }" > new.json && \
+              jq < result.json --argjson new "$(cat new.json)" '. += [$new]' > result.json.tmp && \
+              rm new.json && \
+              mv result.json.tmp result.json
           done
+          DIFFER='def post_recurse(f): def r: (f | select(. != null) | r), .; r; def post_recurse: post_recurse(.[]?); (. | (post_recurse | arrays) |= sort)'
+          diff <(jq -S "$DIFFER" results-main/result.json) <(jq -S "$DIFFER" result.json) || true
+          diff -Baur results-main/result.json result.json || true
           {
             echo "### Failed Interop Tests"
+            SHA=$(cat results-main/baseline-sha.txt || true)
+            if [ -n "$SHA" ]; then
+              {
+                echo "Interop failures relative to $SHA."
+                echo
+              } >> results.md
+            fi
+            if [ -e failed.md ]; then
+              echo "[QUIC Interop Runner](https://github.com/quic-interop/quic-interop-runner), *client* vs. *server*"
+              cat failed.md
+            else
+              echo "None :tada:"
+            fi
+            echo "<details><summary>All results</summary>"
+            echo
             echo "[QUIC Interop Runner](https://github.com/quic-interop/quic-interop-runner), *client* vs. *server*"
-            cat failed.md
-            echo "<details><summary>Succeeded and unsupported tests</summary>"
             for GROUP in succeeded unsupported; do
-              echo
               echo "### ${GROUP^} Interop Tests"
-              echo
-              cat "$GROUP.md"
-              echo
+              if [ -e "$GROUP.md" ]; then
+                cat "$GROUP.md"
+              else
+                echo "None :question:"
+              fi
             done
+            echo
             echo "</details>"
           } >> comment.md
 
+      - name: Remember main-branch push URL
+        # if: github.ref == 'refs/heads/main'
+        run: echo "${{ github.sha }}" > baseline-sha.txt
+
+      - name: Cache main-branch results
+        # if: github.ref == 'refs/heads/main'
+        uses: actions/cache/save@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
+        with:
+          path: |
+            result.json
+            baseline-sha.txt
+          key: qns-${{ runner.os }}-${{ github.sha }}
```

Comment:
Is this right?  Anything that reads the cache will ignore the sha.  I guess that you need unique keys and this is how you achieve that...

Reply:
Yeah, the download step looks like
```
      - name: Download cached main-branch results
        uses: actions/cache/restore@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
        with:
          path: results-main
          key: qns-${{ runner.os }}-${{ github.sha }}
          restore-keys: qns-${{ runner.os }}-
```
and will never match the `key` and always fall back to the `restore-key`.